cmd/sourced/dir/dir.go

```diff
@@ -13,27 +12,69 @@ import (
 	goerrors "gopkg.in/src-d/go-errors.v1"
 )
 
-// ErrNotExist is returned when .sourced dir does not exists
+// ErrNotExist is returned when config dir does not exists
 var ErrNotExist = goerrors.NewKind("%s does not exist")
 
+// ErrNotDirectory is returned when config dir is not a directory
+var ErrNotDirectory = goerrors.NewKind("%s is not a directory")
+
 // Path returns the absolute path for $SOURCED_DIR, or $HOME/.sourced if unset
+// and returns an error if it does not exist or it could not be read.
 func Path() (string, error) {
+	srcdDir, err := path()
+	if err != nil {
+		return "", err
+	}
+
+	info, err := os.Lstat(srcdDir)
+	switch {
+	case os.IsNotExist(err):
+		return "", ErrNotExist.New(srcdDir)
+	case !info.IsDir():
+		return "", ErrNotDirectory.New(srcdDir)
+	case err != nil:
+		return "", err
+	}
+
+	return srcdDir, nil
+}
+
+func path() (string, error) {
 	if d := os.Getenv("SOURCED_DIR"); d != "" {
-		return filepath.Abs(d)
+		abs, err := filepath.Abs(d)
+		if err != nil {
+			return "", errors.Wrap(err, fmt.Sprintf("could not resolve SOURCED_DIR='%s'", d))
+		}
+
+		return abs, nil
 	}
 
 	homedir, err := os.UserHomeDir()
 	if err != nil {
 		return "", errors.Wrap(err, "could not detect home directory")
 	}
 
-	srcdDir := filepath.Join(homedir, ".sourced")
-	_, err = os.Lstat(srcdDir)
+	return filepath.Join(homedir, ".sourced"), nil
+}
+
+// Prepare tries to create the config directory, returning an error if it could not
+// be created, or nill if already exist or was successfully created.
+func Prepare() error {
+	srcdDir, err := path()
+	if err != nil {
+		return err
+	}
+
+	info, err := os.Stat(srcdDir)
 	if os.IsNotExist(err) {
-		return "", ErrNotExist.New(srcdDir)
+		return os.MkdirAll(srcdDir, os.ModePerm)
 	}
 
-	return srcdDir, nil
+	if err != nil || !info.IsDir() {
+		return ErrNotDirectory.New(srcdDir)
```

Comment:
it can be a directory but `os.Stat` can fail to read it

Reply:
True, what if this error is replaced by:
```go
var ErrNotWritableDirectory = goerrors.NewKind("%s is not a writable directory")
```
so it'll describe when `~/.sourced`:
- is not a directory,
- is not writable
- or `os.Stat` can not read it.

Reply:
It can be writable but not readable.

Reply:
It's great to know that with this PR we're identifying and fixing a lot of corner-cases that we didn't identify in the past. :sparkles:  team sport!  :sparkles: 

`s/ErrNotWritableDirectory/ErrNotValidDirectory`, wdyt? suggestions are welcome.

Reply:
I see you are currently hiding the real error when `ErrNotDirectory` happens.
So we have 3 options:
1. Create custom errors for each possible error and continue hiding original err
2. Special error type for example `dir.ErrNotValid` (which still contains the original error as is or wrapped into nicer text because we need to let a user know what went wrong so they can fix it)
2. Just propagate errors as we did before and let main handler to catch them, displaying: `Unexpected error: <error from stdlib about permissions or whatever else>`.

Reply:
IMHO if we have to apply option 1 where we have to create a custom error for each possible original error, then it's better to not create any and use original ones. But original ones could be not friendly enough, so I'd choose option 2.

Reply:
I agree with you @se7entyse7en about choosing `2` from @smacker alternatives
Done by bb1902a